rasa_core/actions/forms.py

```diff
@@ -0,0 +1,138 @@
+        else:
+            return []
+
+
+class BooleanFormField(FormField):
+    """A form field that prompts the user for a yes or no answer.
+    The interpreter should map positive and negative answers to
+    the intents ``affirm_intent`` and ``deny_intent``.
+    """
+    def __init__(self, slot_name, affirm_intent, deny_intent):
+        self.slot_name = slot_name
+        self.affirm_intent = affirm_intent
+        self.deny_intent = deny_intent
+
+    def extract(self, tracker):
+
+        intent = tracker.latest_message.intent.get("name")
+        if intent == self.affirm_intent:
+            value = True
+        elif intent == self.deny_intent:
+            value = False
+        else:
+            return []
+
+        return [SlotSet(self.slot_name, value)]
+
+
+class FreeTextFormField(FormField):
+
+    def __init__(self, slot_name):
+        self.slot_name = slot_name
+
+    def extract(self, tracker):
+        validated = self.validate(tracker.latest_message.text)
+        if validated is not None:
+            return [SlotSet(self.slot_name, validated)]
+        return []
+
+
+class FormAction(Action):
+
+    RANDOMIZE = True
+
+    @staticmethod
+    def required_fields():
+        return []
+
+    def should_request_slot(self, tracker, slot_name, events):
+        existing_val = tracker.get_slot(slot_name)
+        pending = [e.key for e in events if e.key == slot_name]
+        return existing_val is None and slot_name not in pending
+
+    def get_other_slots(self, tracker):
+        requested_slot = tracker.get_slot("requested_slot")
+        slot_events = [f.extract(tracker) for f in self.required_fields()
+                       if isinstance(f, EntityFormField) and
+                       not f.slot_name == requested_slot]
+        return [e for events in slot_events for e in events]
+
+    def get_requested_slot(self, tracker):
+        requested_slot = tracker.get_slot("requested_slot")
+
+        required = self.required_fields()
+
+        if self.RANDOMIZE:
+            random.shuffle(required)
+
+        if requested_slot is None:
+            return []
+        else:
+            fields = [f for f in required if f.slot_name == requested_slot]
+            if len(fields) == 1:
+                return fields[0].extract(tracker)
+            else:
+                logger.debug("unable to extract value "
+                             "for requested slot: {}".format(requested_slot))
+                return []
+
+    def run(self, dispatcher, tracker, domain):
+
+        events = self.get_requested_slot(tracker) + self.get_other_slots(tracker)
+
+        for field in self.required_fields():
+            if self.should_request_slot(tracker, field.slot_name, events):
+                dispatcher.utter_template(
+                    "utter_ask_{}".format(field.slot_name))
+
+                events.append(SlotSet("requested_slot", field.slot_name))
+                return events
+
+        submit_tracker = tracker.copy()
```

Comment:
if we move this up before the `for` we don't need to do the `pending` stuff in `should_request_slot` anymore